### ISSUE TITLE
Unix socket created with wrong permissions

### ISSUE THREAD

**Issue by oddlama** [oauth2-proxy/oauth2-proxy#2141]
## Expected Behavior

Passing `--http-address=unix:///run/oauth2_proxy/oauth2_proxy.sock` will create a file with default permissions, so usually `755` with a 027 umask or `700` with umask 077. Generally, the socket permissions should be independent of the process umask. Ideally it should just default to `770` regardless of umask to allow other processes in the same group to write to it, and be customizable via a cli parameter.

## Current Behavior

The socket will have permissions depending on the process umask, which should not affect this.

**Comment by oddlama** [comment]
Still relevant.

**Comment by JoelSpeed** [comment]
Is this causing issues currently or is it just a security note? How hard would it be to implement this change? Can you raise a PR?

**Comment by oddlama** [comment]
Not strictly a security issue. To connect to a unix socket, you need write permission on the socket. So the current default permissions of `755` are to strict to allow other users (mostly a reverse proxy like nginx) to connect.
Since the unix socket is not forced to be created with 770, the current behaviour encourages setting a more permissive umask like `Umask=007` on the service definition to allow other services to use the socket. But this also allows them to access the internal files that oauth2-service writes to state directory.

So instead, oauth2-proxy should just manually allow users to set the mode of the resulting socket (and default to 770) so that we can keep using a secure umask.

The fix is simple: Add an option `--unix-socket-mode=`, let it default to 770 and call `os.Chmod` on the socket after creating it. Unfortunately I don't have enough time right now to make a PR, but based on my understanding as an outsider this should be basically trivial to implement.

**Comment by JoelSpeed** [comment]
> But this also allows them to access the internal files that oauth2-service writes to state directory.

What files are you referring to? To my knowledge, OAuth2 Proxy writes no files to disk. If the umask solution works and we don't have any other files, I'm not sure there really is an issue.

Though you're right, adding a socket mode would probably be useful to make a complete feature

**Comment by oddlama** [comment]
> > But this also allows them to access the internal files that oauth2-service writes to state directory.
> 
> What files are you referring to? To my knowledge, OAuth2 Proxy writes no files to disk. If the umask solution works and we don't have any other files, I'm not sure there really is an issue.

You are right, I wrongly assumed it would write additional state. While the umask solution works, it's at best a very hacky way to work around the problem.

Setting the umask on the service will affect more than just the service, for example all of the specified pre and post start scripts are also affected. Those will probably not expect a permissive umask so that can lead to more issues. It's kind of similar to setting a file world-readable just to avoid dealing with permission management, it works to solve the problem but could have unintended side effects..

> Though you're right, adding a socket mode would probably be useful to make a complete feature

Since it should be a very simple addition I hope someone can spare the time to implement it at some point in the future.

**Comment by btrepp** [comment]
I've hit this trying to use oauth2-proxy as a sidecar with Nginx rootless to secure an app.
I am volume mounting and sharing a socket fine, but the different containers (Nginx and oauth2-proxy) are ultimately running as different non-root users. So Nginx isn't able to read the socket.

I'm not sure if it's even easy to set a umask on a Kubernetes deployment. It probably will require patching in a custom entry point before the oauth2-proxy starts.

Even if I get them in the same group, I don't think 755 will allow another process to interact with the socket at all. 
From research execute doesn't make sense on sockets swell, so the permissions should really be 0660, allowing the sockets user or group to read and write from it.

I suspect most people are using tcp ports, which really are more 666, there's not security at all on tcp :).

Having this as a flag would be fantastic, that way it could be backwards compatible (keep the current setting), but people can set the permissions as needed.

**Comment by tuunit** [comment]
Have you tried to set the same values for the pod security context for nginx and oauth2-proxy.

Something like:
```yaml
securityContext:
    runAsUser: 1000
    runAsGroup: 3000
    fsGroup: 2000
```

**Comment by tuunit** [comment]
In Linux the user IDs aren't unique to the container. User id 0 is the root user independent of the container or the host and creating a file on a pvc with user id 1000 in one container will grant a user with user id 1000 in another container access to those files when mounting the PVC. Independent of the Linux distro or internal username.

**Comment by btrepp** [comment]
Oh yeah if all containers ran under the same user id I am sure it would work, but, usually they don't.
Eg Nginx by default uses 101, oauth2-proxy looks like it's using 1001. 

Changing all the security contexts to the same id could work, but then that also can cause problems with images. Eg they set up all random files with the inbuilt users (101/1001 etc) and then it all falls over as you are running as 1000 now.

In that case it probably come down to having to patch the other containers to use similar user IDs, so either way I am patching things. 

Creating a custom entry point and patching the umask does work
```
#!/bin/sh
umask 007
/bin/oauth2-proxy $@
```

```
      - name: oauth2-proxy
        image: quay.io/oauth2-proxy/oauth2-proxy
        args: 
        - --config 
        - /etc/oauth2-proxy.cfg
        command:
        - /bin/oauth2-proxy-umask
       .....
       volumeMounts:
          - name: oauth2-proxy
            mountPath: /etc/oauth2-proxy.cfg
            subPath: oauth2-proxy.cfg
          - name: oauth2-proxy-patch
            mountPath: /bin/oauth2-proxy-umask
            subPath: oauth2-proxy-umask.sh
       ....
```


with 
```
  template:
    spec:
      securityContext:
        fsGroup: 2000
```
Which is similar to what the service managers would be doing, so users can work-around this.

I think using security contexts for a common group for the pod, plus having the

**Comment by btrepp** [comment]
Still would be value to do from a simplification/principle of least surprise perspective

**Comment by oddlama** [comment]
Still relevant

**Comment by oddlama** [comment]
Still relevant. Dear actions bot, this isn't going to solve itself.

**Comment by btrepp** [comment]
![93yeye.jpg](https://github.com/user-attachments/assets/34202c84-785e-4d0a-ae9d-6b03d6351061)

**Comment by JoelSpeed** [comment]
Is this a duplicate of/related to https://github.com/oauth2-proxy/oauth2-proxy/issues/2536?

**Comment by oddlama** [comment]
Yes, although this issue is 9 months older so i guess technically #2536 is a duplicate of this one.

**Comment by JoelSpeed** [comment]
Ack, fair point, just trying to make a link since #2537 is supposed to fix #2536. Anyone able to test the proposed fix in their use case?

**Comment by oddlama** [comment]
The fix will definitely work, so you can add this one to the list of issues the PR will close.

**Comment by oddlama** [comment]
Ok i give up.

**Comment by jfly** [comment]
Still relevant
